scripts/source-maintenance.sh

```diff
@@ -40,6 +40,26 @@ fi
 COPYRIGHT_YEARS=`date +"1996-%Y"`
 echo "s/1996-2[0-9]+ The Squid Software Foundation and contributors/${COPYRIGHT_YEARS} The Squid Software Foundation and contributors/g" >>boilerplate_fix.sed
 
+updateifchanged()
+{
+original="$1"
+updated="$2"
+message="$3"
+
+if test -e "${original}" -a -e "${updated}"; then
+	md51=`cat "${original}" | tr -d "\n \t\r" | ${MD5}`;
+	md52=`cat "${updated}" | tr -d "\n \t\r" | ${MD5}`;
+	if test "${md51}" != "${md52}" ; then
+		echo "NOTICE: File ${original} changed: ${message}"
+		mv "${updated}" "${original}"
```

Comment:
The actions in the main script are very modular and error in one place do not preclude following sections from being run. So I will not be adding -e at this time, nor likely for a while.

Reply:
> The actions in the main script are very modular

I agree.

> and error in one place do not preclude following sections from being run. 

IMO, we should not keep modifying the same file after the first modification has failed. 

Yes, every modification is a separate "module", but that does not make dealing with cascading errors simple enough. We could build a blacklist of failed files to preclude _their_ further modification, but it is much better, overall, to stop on the first failure instead IMO, especially when some files may depend on others.

[rant] Imagine a preprocessor that gives the compiler a known-to-be-broken preprocessed source file (because the problem will not be visible to the compiler) and the compiler that discovers more errors but gives the linker that known-to-be-broken object file just because the linker would be able to successfully link that object into an executable. The resulting executable is completely not what the developer wanted, but it "runs"! In some cases, an experienced developer would quickly find the preprocessor error and correct it, but in many other cases, hours or days will be wasted trying to understand why the newly deployed program version brought the server down... [/rant]

> So I will not be adding -e at this time, nor likely for a while.

I do not think I should insist on adding a global `set -e` at this point, especially if there are valid reasons why it would be difficult to add (none known to me at this point, but they could exist).

However, I do insist that the new code checks for errors and quits on the first one. If you find that the best way to accomplish that is a (local) `set -e`, great, but there may be better solutions out there, and I am always open to finding the best solution.

Reply:
IMO the plugin maintenance scripts should never depend on each other. So we should be able to just skip any which goes wrong and nothing will break.

Reply:
> IMO the plugin maintenance scripts should never depend on each other. 

I agree. `Never` might be too strong of a word here (there might be some reasonable exceptions), but it is a valid design principle even if we do allow exceptions, so we do not need to agree on the `never` bit as far as this change request is concerned.

> So we should be able to just skip any which goes wrong and nothing will break.

I see no logical connection between the "independence of plugins" principle and the desire to pile modifications on top of modifications that went wrong. "Nothing will break" is clearly false -- something has gone wrong already. The question in this discussion is not whether something breaks (that is a necessary _precondition_) but whether it is a good idea to pile up modifications on top of failures.

The current PR already contains some improvements in this area. In hope to make progress, I am going to resolve this "outdated" change request and focus on error handing of specific commands in the updated script. Perhaps we can find a way to agree on specific modifications (to continue running after some failures) even if we cannot agree on whether the script should quit on the first failure.